### PR TITLE
Nix2 on 17.09

### PR DESCRIPTION
**Pull request by obadz**
###### Motivation for this change

Wanted to play with Nix 2.0 on 17.09. In case anybody else wants to, just need to add this to your NixOS `configuration.nix`

```nix
{
  nix.package = pkgs.nixStable2;
}
```

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/tools/package-management/nix/default.nix [hunk 1]
```diff
@@ -39,23 +39,23 @@ let
     inherit name src;
     version = lib.getVersion name;
 
-    is112 = lib.versionAtLeast version "1.12pre";
+    is20 = lib.versionAtLeast version "2.0pre";
```
Comment (zimbatm) [inline pkgs/tools/package-management/nix/default.nix]: I'm not sure if this needs to be changed. 2.0 is bigger than 1.12 right?

File: nixos/modules/services/misc/nix-daemon.nix [hunk 2]
```diff
@@ -395,31 +404,26 @@ in
       };
 
     nix.envVars =
-      { NIX_CONF_DIR = "/etc/nix";
-      }
+      optionalAttrs (!isNix20) {
+        NIX_CONF_DIR = "/etc/nix";
```
Comment (zimbatm) [inline nixos/modules/services/misc/nix-daemon.nix]: why is this optional in nix 2.0?

File: nixos/modules/services/misc/nix-daemon.nix [hunk 3]
```diff
@@ -395,31 +404,26 @@ in
...
       }
 
-      // optionalAttrs cfg.distributedBuilds {
-        NIX_BUILD_HOOK =
-          if isNix112 then
-            "${nix}/libexec/nix/build-remote"
-          else
-            "${nix}/libexec/nix/build-remote.pl";
+      // optionalAttrs (cfg.distributedBuilds && !isNix20) {
+        NIX_BUILD_HOOK = "${nix}/libexec/nix/build-remote.pl";
```
Comment (zimbatm) [inline nixos/modules/services/misc/nix-daemon.nix]: is the build hook now inferred?

File: pkgs/tools/package-management/nix/default.nix (hunk 1 above)
Comment (obadz) [reply pkgs/tools/package-management/nix/default.nix]: This is a commit I merely cherry-picked from master :)

**Pull request merged**